src/ILCompiler.ReadyToRun/src/CodeGen/ReadyToRunObjectWriter.cs

```diff
@@ -56,6 +60,34 @@ public void EmitPortableExecutable()
                 _headerSectionIndex = sectionBuilder.AddSection(R2RPEBuilder.TextSectionName, SectionCharacteristics.ContainsCode | SectionCharacteristics.MemExecute | SectionCharacteristics.MemRead, 512);
                 sectionBuilder.SetReadyToRunHeaderTable(_nodeFactory.CoreCLRReadyToRunHeader, _nodeFactory.CoreCLRReadyToRunHeader.GetData(_nodeFactory).Data.Length);
 
+                foreach (var depNode in _nodes)
+                {
+                    if (depNode is MethodCodeNode methodNode)
+                    {
+                        int methodIndex = _nodeFactory.CoreCLRReadyToRunRuntimeFunctionsTable.Add(methodNode);
+                        if (methodNode.Method is EcmaMethod ecmaMethod)
+                        {
+                            // Strip away the token type bits, keep just the low 24 bits RID
+                            int rid = MetadataTokens.GetToken(ecmaMethod.Handle) & 0x00FFFFFF;
+                            Debug.Assert(rid != 0);
+                            
+                            // TODO: how to synthesize method fixups blob?
+                            byte[] fixups = null;
+                            _nodeFactory.CoreCLRReadyToRunMethodEntryPointTable.Add(rid - 1, methodIndex, fixups, signature: null, methodHashCode: 0);
+                        }
+
+                        // TODO: method instance table
+                    }
+                    if (depNode is EETypeNode eeTypeNode &&
+                        eeTypeNode.Type is EcmaType ecmaType)
+                    {
+                        int rid = MetadataTokens.GetToken(ecmaType.Handle) & 0x00FFFFFF;
```

Comment:
Nit: if I was doing this, I would try to put all the data structure details (the fact we store a RID) in the table class itself. It would keep the format details in a single spot instead of dragging it into multiple files (i.e. I would only do `_nodeFactory.CoreCLRReadyToRunTypesTable.Add(eeTypeNode);` and have the table extract the rid when it needs it. Doing that also saves memory.)

Reply:
I can easily refactor it like that - in fact, the main reason why I kept it separate in the first version was that I was unsure about the logic involved. In particular, I either need to check whether a method or type is Ecma upfront - but then it seems wasteful to do the casting a second time in the R2R table builder, or alternatively I can just pass the method / type node to the builder - but then I don't know what the builder should do when hitting a method / type which is not Ecma. Perhaps it should just throw a non-implemented exception, that sounds like a good 1st iteration; for now I assume I'll rework it like this unless I hear from you otherwise.